galaxyproject/galaxy
# Update GPU Jupyterlab interactive tool

This PR adds two features to the GPU-enabled JupyterLab interactive tool:

1. Run as a tool. This feature has been inspired by Jupytool (https://github.com/galaxyproject/galaxy/pull/13606). This GPU-enabled Jupyterlab can also be used as a tool and it can execute an Ipython script (in `previous` mode) and create datasets in a Galaxy history similar to many other Galaxy tools in addition to opening several default notebooks when run in `scratch` mode.

2. `Root` user has been removed from the Docker container (https://github.com/anuprulez/ml-jupyter-notebook/blob/resize/Dockerfile) as well as from the tool's definition. To provide permissions to the custom notebook user to Galaxy's datasets, `$default_file_path` parameter in job_conf.xml should be set to read-write (`$default_file_path:rw`). Alternatively, to assign necessary permissions to this custom user, the properties `docker_volumes` from job_conf.xml and `outputs_to_working_directory` in galaxy.yml should not be set.

## How to test the changes?
(Select all options that apply)
- [x ] Instructions for manual testing are as follows:
  1. Bring Galaxy's config (galaxy.yml) and job_conf.yml config changes used for running interactive tools (galaxy.yml.interactivetools and job_conf.xml.interactivetools). In addition, in "docker_volumes" of job_conf.xml, the permission of `$default_file_path:rw` should be set as read-write. Since, the notebook's container and associated script run with custom user's permissions, this user does not have permission to use Galaxy's datasets.
  2. Include this tool in the Galaxy's default list of tools (configure in tool conf)

## License

![gpu_jupyterlab_1](https://user-images.githubusercontent.com/3022518/204488929-6d914744-755c-41e1-838b-cc45004c182a.png)

![gpu_jupyterlab_2](https://user-images.githubusercontent.com/3022518/204488927-71b95936-7502-4c77-a0df-27910991f891.png)

![gpu_jupyterlab_3](https://user-images.githubusercontent.com/3022518/204488919-0cc94208-ad1e-4132-87b4-66f41758bddc.png)

![gpu_jupyterlab_4](https://user-images.githubusercontent.com/3022518/204488925-95b33746-9ce5-43c7-8c7b-9bd3f3e17366.png)

![gpu_jupyterlab_5](https://user-images.githubusercontent.com/3022518/204489649-eee0ff01-50af-4cb3-b114-6b17c92ee942.png)
